File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:

    def check(self,x:int,nums:List[int],target:int)->bool:
        return (nums[x]-target)>=0
    def search(self, nums: List[int], target: int) -> int:
        n=len(nums)
        lo=0
        hi=n-1
        ans=nums[0]
        while lo<hi:
            mid=lo+(hi-lo)//2
            if nums[mid]==target:
                return mid
            else:
                if self.check(mid,nums,target)==1:
                    hi=mid
                    ans=mid
                else:
                    lo=mid+1
        if nums[ans]==target:
            return ans
        else:
            return -1
        

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
bool check(vector<int> &nums,int i){
    if(nums[i]-nums[0]>=0) return true;
return false;
}
    int search(vector<int>& nums, int target) {
        int n=nums.size();
        int low=0;
        int high=n-1;
        int ans=0;
        while(low<=high){
            int mid=low+(high-low)/2;
if(check(nums,mid)){
    low=mid+1;
}
else{
    ans=mid;
    high=mid-1;
}
        }

int pivot=ans;

if(nums[0]<=target){
    low=0;high=pivot-1;
}
else{
    low=pivot;high=n-1;
}

int final_ans=-1;
while(low<=high){
    int mid=low+(high-low)/2;
if(check(nums,mid)){
    low=mid+1;
}
else{
    final_ans=mid;
    high=mid-1;
}
}
if(nums[final_ans]==target)
return final_ans;
else return -1;
    }
};
